pkg/sidecar/sidecar.go

```diff
@@ -84,16 +89,48 @@ func Run(argv []string, version string) int {
 
 	klog.Infof("CSI driver name: %q", rt.DriverName)
 
-	// TBD May need to exposed metric HTTP end point
-	// here because the wait for the CSI driver is open ended.
+	rt.MetricsManager = metrics.NewMetricsManager()
+	wg := &sync.WaitGroup{}
+
+	mux := http.NewServeMux()
+	if *s.httpEndpoint != "" {
+		err := rt.MetricsManager.PrepareMetricsPath(mux, *s.metricsPath, promklog{})
+		if err != nil {
+			klog.Errorf("Failed to prepare metrics path: %s", err.Error())
+			os.Exit(1)
+		}
+		klog.Infof("Metrics path successfully registered at %s", *s.metricsPath)
+	}
 
 	grpcServer, err := startGRPCServerAndValidateCSIDriver(rt)
 	if err != nil {
 		klog.Error(err)
 		return 1
 	}
 
-	// TODO: Start the HTTP metrics server here.
+	// start listening & serving http endpoint if set
+	if *s.httpEndpoint != "" {
```

Comment:
Are there any unit tests you can add to this package to ensure that server is launched correctly?

Reply:
I have added the unit test [here](https://github.com/kubernetes-csi/external-snapshot-metadata/blob/1d3a2a002aadf18f74673dffcb67df44a6cd6eee/pkg/metrics/metrics_test.go#L63-L82) that tests the successful start of the server.

Reply:
Its not the same thing as it is not testing _this_ http server's configuration.

I suggest splitting the new logic into 2 functions that can be unit tested in this package itself, by, perhaps, making a client call to the server.  e..g.
```
...
if *s.httpEndpoint != "" {
     srv := startMetricsServer(*s.httpEndPoint, *s.metricsPath, rt)
     defer stopMetricsServer(srv)
}

grpcServer, err := startGRPCServerAndValidateCSIDriver(s.createServerConfig(rt))
...
```
where `startMetricsServer` looks something like this:
```
func startMetricsServer(httpEndpoint string, metricsPath string, rt *runtime.Runtime) *http.Server {
	wg := &sync.WaitGroup{}
        wg.Add(1) // <--- missing in the original logic

	mux := http.NewServeMux()
	err := rt.MetricsManager.PrepareMetricsPath(mux, metricsPath, promklog{}) // <--- try to remove promklog{}
	if err != nil {
		klog.Errorf("Failed to prepare metrics path: %s", err.Error())
		os.Exit(1)
	}
	klog.Infof("Metrics path successfully registered at %s", metricsPath)

	l, err := net.Listen("tcp", httpEndpoint)
	if err != nil {
		klog.Fatalf("failed to listen on address[%s], error[%v]", httpEndpoint, err)
	}

	srv := &http.Server{Addr: l.Addr().String(), Handler: mux}
	go func() {
		wg.Done()
		if err := srv.Serve(l); err != http.ErrServerClosed {
			klog.Fatalf("failed to start endpoint at:%s/%s, error: %v", httpEndpoint, metricsPath, err)
		}
	}()

	wg.Wait() // <--- missing in the original logic
	klog.Infof("Metrics http server successfully started on %s, %s", httpEndpoint, metricsPath)
}
```
and `stopMetricsServer` looks something like this:
```
func stopMetricsServer(srv *http.Server) {
	if err := srv.Shutdown(context.Background()); err != nil {
		klog.Errorf("Failed to shutdown metrics server: %s", err.Error())
	}
	klog.Infof("Metrics server successfully shutdown")
}
```

Reply:
Understood, this makes sense.
Though, there is one catch that we are starting the http server even before the sidecar GRPC server is validated and started, which in best cases shouldn't be a problem but logically we shouldn't be doing, right?
We should start the httpServer only post the validation and the the sidecar server is started.
Let's further split the function by moving the `PrepareMetricsPath` code before the `startGRPCServerAndValidate` call, and then start the http server. WDYT?
Like this:
```
	mux := http.NewServeMux()
	if *s.httpEndpoint != "" {
		err := rt.MetricsManager.PrepareMetricsPath(mux, *s.metricsPath)
		if err != nil {
			klog.Errorf("Failed to prepare metrics path: %s", err.Error())
			os.Exit(1)
		}
		klog.Infof("Metrics path successfully registered at %s", *s.metricsPath)
	}

	grpcServer, err := startGRPCServerAndValidateCSIDriver(rt)
	if err != nil {
		klog.Error(err)
		return 1
	}

	// start listening & serving http endpoint if set
	if *s.httpEndpoint != "" {
		srv := startMetricsServer(*s.httpEndpoint, *s.metricsPath, mux, rt)
		if srv != nil {
			defer stopMetricsServer(srv)
		}
	}

	shutdownOnTerminationSignal(grpcServer)

	return 0
}

func startMetricsServer(httpEndpoint string, metricsPath string, mux *http.ServeMux, rt *runtime.Runtime) *http.Server {
	wg := &sync.WaitGroup{}
	wg.Add(1)

	l, err := net.Listen("tcp", httpEndpoint)
	if err != nil {
		klog.Fatalf("failed to listen on address[%s], error[%v]", httpEndpoint, err)
		return nil
	}

	srv := &http.Server{Addr: l.Addr().String(), Handler: mux}
	go func() {
		wg.Done()
		if err := srv.Serve(l); err != http.ErrServerClosed {
			klog.Fatalf("failed to start endpoint at:%s/%s, error: %v", httpEndpoint, metricsPath, err)
		}
	}()

	wg.Wait()
	klog.Infof("Metrics http server successfully started on %s, %s", httpEndpoint, metricsPath)

	return srv
}

func stopMetricsServer(srv *http.Server) {
	if err := srv.Shutdown(context.Background()); err != nil {
		klog.Errorf("Failed to shutdown metrics server: %s", err.Error())
	}
	klog.Infof("Metrics server successfully shutdown")
}
```

Reply:
I've no objection - you have a better understanding of what is needed.